sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -137,8 +137,7 @@ class CodeGenContext {
     dataType match {
       case _ if isPrimitiveType(jt) => s"$row.set${primitiveTypeName(jt)}($ordinal, $value)"
       case t: DecimalType => s"$row.setDecimal($ordinal, $value, ${t.precision})"
-      // The UTF8String may came from UnsafeRow, otherwise clone is cheap (re-use the bytes)
-      case StringType => s"$row.update($ordinal, $value.clone())"
+      case StringType => s"$row.update($ordinal, $value)"
```

Comment:
I moved the `copy` [here](https://github.com/apache/spark/pull/8029/files#diff-36fe7bbf96798a372dda38271e32103dR125), just want to tighten up the scope that we only need to do copy when convert a unsafe row to safe row(according to the fact that we only copy nested struct in `GenerateSafe` not here before).